nautobot/dcim/models/devices.py

```diff
@@ -955,6 +956,55 @@ def get_children(self):
         """
         return Device.objects.filter(parent_bay__device=self.pk)
 
+    @property
+    def modules(self):
+        """
+        Return all child Modules installed in ModuleBays within this Device.
+        """
+        # Supports Device->ModuleBay->Module->ModuleBay->Module->ModuleBay->Module->ModuleBay->Module
+        # This query looks for modules that are installed in a module_bay and attached to this device
+        # We artificially limit the recursion to 4 levels or we would be stuck in an infinite loop.
+        recursion_depth = 4
+        qs = Module.objects.all()
+        query = models.Q()
+        for level in range(recursion_depth):
+            recursive_query = "parent_module_bay__parent_module__" * level
+            query = query | models.Q(**{f"{recursive_query}parent_module_bay__parent_device": self})
+        return qs.filter(query)
+
+    @property
+    def all_console_ports(self):
+        # TODO: These could probably be optimized to reduce the number of joins
+        return ConsolePort.objects.filter(models.Q(device=self) | models.Q(module__in=self.modules))
+
+    @property
+    def all_console_server_ports(self):
+        return ConsoleServerPort.objects.filter(models.Q(device=self) | models.Q(module__in=self.modules))
+
+    @property
+    def all_front_ports(self):
+        return FrontPort.objects.filter(models.Q(device=self) | models.Q(module__in=self.modules))
+
+    @property
+    def all_interfaces(self):
+        return self.vc_interfaces | Interface.objects.filter(module__in=self.modules)
+
+    @property
+    def all_module_bays(self):
+        return ModuleBay.objects.filter(models.Q(parent_device=self) | models.Q(parent_device__in=self.modules))
```

Comment:
Is this one right? `parent_device__in=self.modules` looks sketchy to me.

Reply:
Should be `parent_module__in=self.modules`